wp-content/plugins/tapestry/controller/class.tapestry-controller.php

```diff
@@ -3,4 +3,106 @@
  * Add/update/retrieve Tapestry post and its child nodes
  * 
  */
-class TapestryController {}
\ No newline at end of file
+
+class TapestryController {
+    const POST_TYPES = [
+        'TAPESTRY' => 'tapestry',
+        'TAPESTRY_NODE' => 'tapestry_node'
+    ];
+    const ERRORS = [
+        'INVALID_POST_ID' => [
+            'MESSAGE' => 'PostID is invalid',
+            'STATUS' => ['status' => 404]
+        ]
+    ];
+    private $postId;
+
+    /**
+     * Constructor
+     */
+    public function __construct($postId) {
+        $this->postId = $postId;
+    }
+
+    /**
+     * Update Tapestry nodes first then
+     * Update the existing Tapestry if the postId is provided
+     * Otherwise, a new Tapestry will be created
+     * 
+     * @param  Object @tapestry The Tapestry data
+     * @return Object @tapestry
+     */
+    public function updateTapestry($tapestry) {
+        // TODO: use isValidPostID() utlility function
+        if (!isset($this->postId)) {
+            $this->postId = $this->_updatePost($tapestry, 'tapestry');
+        }
+
+        $this->_updateNodes($tapestry->nodes, $this->postId);
+
+        if (!isset($tapestry->rootId)) {
+            $tapestry->rootId = $tapestry->nodes[0]->id;
```

Comment:
If there are no nodes passed in, this may throw a fatal error.

Reply:
Interestingly, when I tested with no nodes passed it. It doesn't throw any internal server error. It just makes the `$tapestry->rootId` null. But you make a good point, we should check if the nodes array empty or not.